feat: Add new report, and alert type: webhook

kistoth90 opened the pull request:
<!---
Please write the PR title following the conventions at https://www.conventionalcommits.org/en/v1.0.0/
Example:
fix(dashboard): load charts correctly
-->

### SUMMARY
I add **webhook** as a new selectable report type. 

![Képernyőkép ekkor: 2024-08-13 11-24-05](https://github.com/user-attachments/assets/cbadedc9-052c-460f-b05b-3779aaa20069)

By the use of webhook the user can post csv/png/pdf to any endpoint, with some informative parameters as:
- name of the report
- id of the report
- report type (chart, or dashboard)
- content id
- content format

For security reasons I made a new optional parameter: WEBHOOK_SECRET. It's a secret string between the Superset and the webhook endpoint. In the case where WEBHOOK_SECRET is filled, Superset adds an "X-Webhook-Signature" parameter to the header of the post call, which hashes the json data to be sent. The receiving party can verify that the party sending the webhook is the real sender by hashing the received json data and comparing it with the sending party's "X-Webhook-Signature" parameter. If the two parameters do not match, the receiving party may reject the call because the sender is not the supposed host.

![Képernyőkép ekkor: 2024-08-13 12-04-57](https://github.com/user-attachments/assets/99f1dcb2-d739-4c25-b8f5-4678e00717e1)

On the report log, we can get feedback about the successful, and not successful requests as well.

![Képernyőkép ekkor: 2024-08-13 12-27-31](https://github.com/user-attachments/assets/6041296a-07b1-4d0e-958a-8ff73078bac3)

I'm happy to share my code with you guys!

### TESTING INSTRUCTIONS
0. (optional) Fill the WEBHOOK_SECRET input parameter in the config file
1. On the Alerts & reports page, press the "+ Report" or "+ Alert" button.
2. On the notification panel, you can select Webhook as Notification method
3. Fill the webhook endpoint url
4. Fill all the mandatory input 

By the result you will get the selectend chart/dashboard file on your webhook endpoint with some informative parameters as:
- name of the report
- id of the report
- report type (chart, or dashboard)
- content id
- content format

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

Commit:
Add new report, and alert type: webhook
- new report type
- new notification logic
- update NotificationMethod.tsx with webhook
- update config with WEBHOOK_SECRET

villebro commented:
Really cool idea @kistoth90!

Commit:
Fix merge blocking errors:
- add license template to webhook.py
- revert label rename

Commit:
Add pre-commit related fixes, and update NotificationContent with optional id

rusackas commented:
I love the idea here. Approving the next CI run... ping me if it needs to be kicked again.

kistoth90 commented:
At this point I don't know what, or how should I fix the errors:
- I tried to run the cypress on my desk (cypress-run-chrome) but I got constant "400 Bad Request: The CSRF token is missing." during the login call. I was able to log in manually by the test credentials.
- the pre-commit works locally fine... i dont know why it want to refresh the source code during the CI run.

I would appreciate some assistance about, becauseI'm stuck.

dpgaspar commented on superset/reports/notifications/webhook.py:
```diff
@@ -0,0 +1,139 @@
+        """
+        Retrieves the secret token for HMAC signature from the application configuration.
+        """
+        with app.app_context():
```
do we need to create a new app_context here?

dpgaspar commented on superset/reports/notifications/webhook.py:
```diff
@@ -0,0 +1,139 @@
+            return hmac.new(
+                secret.encode("utf-8"),
+                json.dumps(data).encode("utf-8"),
+                digestmod="sha256",
```
digestmod could be a new configuration key

dpgaspar commented on superset/reports/notifications/webhook.py:
```diff
@@ -0,0 +1,139 @@
+        return json.loads(self._recipient.recipient_config_json)["target"]
+
+    def _generate_signature(
+        self, data: Dict[str, Optional[str | None | int]]
```
we still support python 3.9, in order to use the new union type annotation we need to import:
```
from __future__ import annotations
```

dpgaspar commented on superset/reports/notifications/webhook.py:
```diff
@@ -0,0 +1,139 @@
+        return None
+
+    def _get_header(
+        self, data: Dict[str, Optional[str | None | int]]
```
if using `from __future__ import annotations` let's change this to:
``` python
def _get_header(
    self, data: dict[str, Optional[str | None | int]]
...
```

dpgaspar commented on superset/reports/notifications/webhook.py:
```diff
@@ -0,0 +1,139 @@
+            ),
+            "content_format": self._content.header_data.get("notification_format"),
+            "metadata": json.loads(self._recipient.recipient_config_json).get(
+                "JSONData"
```
WDYT about creating a TypeDict to make this payload structure more explicit

dpgaspar commented on superset/reports/notifications/webhook.py:
```diff
@@ -0,0 +1,139 @@
+                timeout=10,
+            )
+            response.raise_for_status()
+            logger.info("Report webhook has been successfully sent to {host}")
```
use:
```
logger.info("Report webhook has been successfully sent to %s", host)
```
instead

fisjac commented:
For the Pre-commit issue, have you tried running ruff and black locally to remove the formatting blocks?

Commit:
fix _construct_payload function: remove content name split

kistoth90 committed:
Merge branch 'apache:master' into dev

Commit:
update config.py

Commit:
fix webhook.py:
- add __future__ package
- remove app_context from _get_secre_token()
- move digestmod to config as constant
- remove metadata from payload
- fix logger.info message

Commit:
Merge branch 'dev' of https://github.com/kistoth90/superset into dev

kistoth90 commented:
Hi @fisjac,

Yepp, the ruff, and Blacklist passed as well!

I made some changes by the guidance of dpgaspar. Let's hope the best! 🤞

sadpandajoe commented:
Closing and reopening to try to get tests to kick off.

zerosnacks commented:
Hi @dpgaspar / @eschutho are there any blockers from merging this?

eschutho commented:
/testenv up FEATURE_ALERT_REPORTS=true

eschutho commented:
It sounds like this feature may be a good use-case for putting behind a feature flag. I can see that some superset administrators may not want to enable webhooks to their users, even if they have the alerts/reports feature on.

eschutho reviewed (changes requested):
Thanks for this contribution! Do you mind writing some tests, or adding to the existing ones with this new functionality? I also left a comment about putting this behind a feature flag.

Shah-Panam commented:
Any update on this? This feature will be highly appreciated.

rita-gorokhod commented:
@kistoth90 would be great if this can be merged, really looking forward to this feature 🚀

Shah-Panam commented:
@kistoth90 @eschutho @sadpandajoe A gentle ping regarding this.

kistoth90 committed:
Merge branch 'master' into webhook

persistanceTube commented:
@kistoth90 hi，is your code suitable for Superset version 4.1.0?

rusackas commented:
Hey @kistoth90 - any chance of getting this rebased so we can help get it across the finish line? I was just asked on [Superset Slack](http://bit.ly/join-superset-slack) if this feature exists, and it sure would be cool if it did! :)

rusackas commented:
Currently the PR needs a rebase, and needs a response to the comments/change requests made by @eschutho. I'm equally excited to see this merged, so I hope it happens!

rusackas commented:
I would love to see the options for a webhook be configurable in the UI, if there's a secure way to do it... maybe treat it like a password, so it doesn't appear in the UI? Then users could add multople webhook configs without having to bother an admin or restarting the service.

rusackas commented:
@kistoth90 can you rebase this PR? I think we're all pretty excited to see it merge :D
